lib/two_factor_authentication/models/two_factor_authenticatable.rb

```diff
@@ -11,17 +11,37 @@ module ClassMethods
         def has_one_time_password(options = {})
           include InstanceMethodsOnActivation
           include EncryptionInstanceMethods if options[:encrypted] == true
-
-          before_create { populate_otp_column }
         end
 
         ::Devise::Models.config(
-          self, :max_login_attempts, :allowed_otp_drift_seconds, :otp_length,
-          :remember_otp_session_for_seconds, :otp_secret_encryption_key)
+          self, :max_login_attempts, :allowed_otp_drift_seconds, :otp_length, :direct_otp_length,
+          :direct_otp_valid_for, :remember_otp_session_for_seconds, :otp_secret_encryption_key)
       end
 
       module InstanceMethodsOnActivation
         def authenticate_otp(code, options = {})
+          return true if direct_otp and authenticate_direct_otp(code, options)
+          return true if totp_enabled? and authenticate_totp(code, options)
+          false
+        end
+
+        def authenticate_direct_otp(code, options = {})
+          return false if direct_otp.nil?
+
+          return false if direct_otp != code
+
+          if Time.now.utc > direct_otp_sent_at + self.class.direct_otp_valid_for
+            # OTP has expired
+            return false
+          end
+
+          # OTP is valid, but can only be used once so clear it from the DB
+          update_columns(direct_otp: nil, direct_otp_sent_at: nil)
+          true
+        end
+
+        def authenticate_totp(code, options = {})
+          raise "authenticate_totp called with no otp_secret_key set" if otp_secret_key.nil?
```

Comment:
I think its useful to users to be able to call authenticate_totp or authenticate_direct_otp directly rather than the hybrid function that basically allows for both.   Some application might want to be clear about exactly what type of code the user is trying to use (e.g. Google does this.. it won't allow you to type a direct OTP in into the TOTP dialog box).